query/src/main/scala/filodb/query/exec/SetOperatorExec.scala

```diff
@@ -187,59 +188,175 @@ final case class SetOperatorExec(queryContext: QueryContext,
   // scalastyle:on method.length
 
 
+  //scalastyle:off method.length
   /***
    * Add all LHS range vector in result. From RHS only only add the range vectors which have join key not
    * present in result
    */
   private def setOpOr(lhsRvs: List[RangeVector]
                       , rhsRvs: List[RangeVector]): Iterator[RangeVector] = {
 
-    val lhsResult = new mutable.HashMap[Map[Utf8Str, Utf8Str], ArrayBuffer[RangeVector]]()
-    val rhsResult = new mutable.HashMap[Map[Utf8Str, Utf8Str], ArrayBuffer[RangeVector]]()
+    val lhsResult = groupAndStitchRangeVectors(lhsRvs)
+    val rhsResult = groupAndStitchRangeVectors(rhsRvs)
+    // All lhsResults should be considered
+    // All rhsResults whose joinKey is not present in lhsResults should be considered
+    // While iterating all lhs rangevectors for a time instant where none of the lhs range vectors are having a value
+    // all RHS range vectors for that time instant having a matching join key will have a value in o/p
+
+    //lhsResult.valuesIterator.map(_.toIterator).flatten ++ rhsResult.valuesIterator.map(_.toIterator).flatten
+
+    // Create a Map of join key as key and value is a two tuple where _1 is all LHS Rvs for the join key and _2
+    // are all RHS Rvs for that same join key
+    val lRvs = lhsResult.foldLeft(Map.empty[Map[Utf8Str, Utf8Str], (Seq[RangeVector], Seq[RangeVector])]) {
+      case (acc, (jk, rvs)) => acc + (jk -> ((rvs, Seq.empty[RangeVector])))
+    }
+
+    val joinedRvs = rhsResult.foldLeft(lRvs) {
+      case (acc, (jk, rvs))  =>
+        val newVal = acc.get(jk).map {
+            // RHS will be Nil
+          case (lhs, rhs) => (lhs, rvs)
+        }.getOrElse((Nil, rvs))
+        acc + (jk -> newVal)
+    }
+
+    case class OrSetOpIteratorWrapper(lhs: Seq[RangeVector], rhs: Seq[RangeVector]) {
+      // Has a side effect where LHS are consumed first and a bit is set to indicate the time instant where there
+      // is at least one non Nan value. the rhs values are then iterated over and emitted only where the given timestep
+      // has no value in any of the LHS timeseries
+      assert(outputRvRange.isDefined, "Expected to see a non null outputRvRange")
+      val (start, end, step) = (outputRvRange.get.startMs, outputRvRange.get.endMs, outputRvRange.get.stepMs)
+      // 64 bits in one word (long), 1 bit per step
+      assert(start == end || step > 0, "Expected to see a non 0 step size when start and end are not same")
+      val wordsRequired = if (start == end) 1 else ((((end - start) / step) >> 6) + 1).toInt
```

Comment:
I'm not  sure any optimization is done behind the scene.

Reply:
Super nitpick: please update this if you see this comment before the merge (I don't see this in the updates since the last review, but it's not worth calling out / blocking the merge).